### PR TITLE
Support Reading Encrypted S3 Objects of Unknown Size

### PR DESCRIPTION
**Pull request by dain**
Slightly different solution to #2434

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java [hunk 1]
```diff
@@ -191,7 +191,9 @@ private static OrcPageSource createOrcPageSource(
         AggregatedMemoryContext systemMemoryUsage = newSimpleAggregatedMemoryContext();
         try {
             OrcReader reader = new OrcReader(orcDataSource, options);
-
+            if (reader.isEmpty()) {
```
Comment (pettyjamesm) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java]: Looks like this would also have to handle handle the case where split start + length exceeds the discovered file size as well as the rarer case where start >= discovered file size.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 2]
```diff
@@ -167,10 +167,11 @@ private static ParquetPageSource createParquetPageSource(
         try {
             FileSystem fileSystem = hdfsEnvironment.getFileSystem(user, path, configuration);
             FSDataInputStream inputStream = hdfsEnvironment.doAs(user, () -> fileSystem.open(path));
-            ParquetMetadata parquetMetadata = MetadataReader.readFooter(inputStream, path, fileSize);
+            dataSource = new HdfsParquetDataSource(new ParquetDataSourceId(path.toString()), estimatedFileSize, inputStream, stats, options);
```
Comment (pettyjamesm) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetPageSourceFactory.java]: This needs to handle start + length > discovered fileSize and start >= discovered file size as well.

File: presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergPageSourceProvider.java [hunk 3]
```diff
@@ -158,10 +159,10 @@ private static ConnectorPageSource createParquetPageSource(
         try {
             FileSystem fileSystem = hdfsEnvironment.getFileSystem(user, path, configuration);
             FileStatus fileStatus = fileSystem.getFileStatus(path);
-            long fileSize = fileStatus.getLen();
+            long estimatedFileSize = fileStatus.getLen();
```
Comment (pettyjamesm) [inline presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergPageSourceProvider.java]: This is an exception because actually the getFileStatus call here will actually be the exact file size now. If this code changes to avoid the status call, then it'll need to handle start >= actual file size and start + length > actual file size just like the other page sources.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/HdfsOrcDataSource.java [hunk 4]
```diff
@@ -54,6 +56,15 @@ public void close()
         inputStream.close();
     }
 
+    @Override
+    public Slice readTail(int length)
+            throws IOException
+    {
+        //  Handle potentially imprecise file lengths by reading the footer
```
Comment (pettyjamesm) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/HdfsOrcDataSource.java]: Should probably record bytes read and timing like readInternal does.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetPageSourceFactory.java (hunk 2 above)
Comment (dain) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/parquet/ParquetPageSourceFactory.java]: I believe the split region (i.e., `start` and `length`) is handled below when `footerBlocks`.  Basically, the code filters the block as declared in the footer to only those that start withing the split region, and only those blocks are processed.  If the split region is off the end of the file it shouldn't matter because no actual blocks can appear in that region.

File: presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergPageSourceProvider.java (hunk 3 above)
Comment (dain) [reply presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergPageSourceProvider.java]: I believe the same reasoning about split region applies to Iceberg.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/HdfsOrcDataSource.java (hunk 4 above)
Comment (dain) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/HdfsOrcDataSource.java]: Yep.  Missed it in the parquet code also.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java (hunk 1 above)
Comment (dain) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java]: ORC uses a similar system to parquet.  The split region is basically a selection for ORC stripes.  Only stripes that start within the split region are considered part of the split, so if the split region is off the end of the file, there will be no blocks in that region.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/rcfile/RcFilePageSourceFactory.java [hunk 5]
```diff
@@ -132,26 +155,32 @@ else if (deserializerClassName.equals(ColumnarSerDe.class.getName())) {
             throw new PrestoException(HIVE_CANNOT_OPEN_SPLIT, splitError(e, path, start, length), e);
         }
 
+        length = min(dataSource.getSize() - start, length);
+        // Split may be empty now that the correct file size is known
+        if (length <= 0) {
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/rcfile/RcFilePageSourceFactory.java]: I didn't follow that.  Can you be more specific?

[the remark this answers is not in the document]
Comment (pettyjamesm) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/rcfile/RcFilePageSourceFactory.java]: On line 128 an empty estimated file size is handled by throwing an exception, but after resolving the actual size we’re just returning an empty page source.

Maybe both situations should just return empty page sources?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java (hunk 1 above)
Comment (pettyjamesm) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcPageSourceFactory.java]: Got it, makes sense that the check isn't necessary for ORC and Parquet then.

**Comment by dain** [comment]
> Still an open question about whether we need to change the handling of empty ORC files before and after exact file size is determined to be consistent with each other, looks good other than that.

I changed `OrcReader` to have a static factory that returns `Optional<OrcReader>`.

**Comment by sopel39** [comment]
Is it meant to read padded files as in: https://en.wikipedia.org/wiki/Padding_(cryptography)#PKCS%235_and_PKCS%237? In such case it's possible to easily figure what is actual content length by reading last `com.amazonaws.services.s3.internal.crypto.JceEncryptionConstants#SYMMETRIC_CIPHER_BLOCK_SIZE` bytes from file.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestOrcPageSourceMemoryTracking.java [hunk 6]
```diff
@@ -194,7 +194,13 @@ private void testPageSource(boolean useCache)
         FileFormatDataSourceStats stats = new FileFormatDataSourceStats();
         ConnectorPageSource pageSource = testPreparer.newPageSource(stats, useCache ? CACHED_SESSION : UNCACHED_SESSION);
 
-        assertEquals(pageSource.getSystemMemoryUsage(), 0);
+        if (useCache) {
```
Comment (electrum) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestOrcPageSourceMemoryTracking.java:197]: Should these be in the previous commit *Add memory tests for both cached and uncached ORC*?

File: presto-orc/src/main/java/io/prestosql/orc/MemoryOrcDataSource.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        requireNonNull(diskRanges, "diskRanges is null");
+
+        if (diskRanges.isEmpty()) {
+            return ImmutableMap.of();
+        }
+
+        ImmutableMap.Builder<K, OrcDataReader> slices = ImmutableMap.builder();
+        for (Entry<K, DiskRange> entry : diskRanges.entrySet()) {
+            DiskRange diskRange = entry.getValue();
+            Slice slice = readFully(toIntExact(diskRange.getOffset()), diskRange.getLength());
```
Comment (electrum) [inline presto-orc/src/main/java/io/prestosql/orc/MemoryOrcDataSource.java:95]: `toIntExact` should not be needed here

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeleteDeltaPageSourceFactory.java [hunk 8]
```diff
@@ -43,9 +46,9 @@ public OrcDeleteDeltaPageSourceFactory(
         this.stats = requireNonNull(stats, "stats is null");
     }
 
-    public OrcDeleteDeltaPageSource createPageSource(Path path, long fileSize)
+    public Optional<ConnectorPageSource> createPageSource(Path path, long fileSize)
     {
-        return new OrcDeleteDeltaPageSource(
+        return OrcDeleteDeltaPageSource.createOrcDeleteDeltaPageSource(
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeleteDeltaPageSourceFactory.java:51]: Static import

File: presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeleteDeltaPageSource.java [hunk 9]
```diff
@@ -96,26 +97,11 @@ public OrcDeleteDeltaPageSource(
...
-                    rowIdColumns,
-                    ImmutableList.of(BIGINT, BIGINT),
-                    OrcPredicate.TRUE,
-                    0,
-                    fileSize,
-                    UTC,
-                    systemMemoryContext,
-                    MAX_BATCH_SIZE,
-                    exception -> handleException(orcDataSource.getId(), exception));
+            Optional<OrcReader> orcReader = OrcReader.createOrcReader(orcDataSource, options);
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeleteDeltaPageSource.java:100]: Static import

File: presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergPageSourceProvider.java [hunk 10]
```diff
@@ -254,7 +254,8 @@ private static ConnectorPageSource createOrcPageSource(
                     inputStream,
                     stats);
 
-            OrcReader reader = new OrcReader(orcDataSource, options);
+            OrcReader reader = OrcReader.createOrcReader(orcDataSource, options)
+                    .orElseThrow(() -> new PrestoException(ICEBERG_MISSING_DATA, "File is empty"));
```
Comment (electrum) [inline presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergPageSourceProvider.java:259]: `MISSING_DATA` is used for HDFS `BlockMissingException`. This is like a corrupt file or something. How about
```java
.orElseThrow(() -> new PrestoException(ICEBERG_BAD_DATA, "ORC file is zero length"));

File: presto-orc/src/main/java/io/prestosql/orc/OrcDataSource.java [hunk 11]
```diff
@@ -29,7 +29,10 @@
 
     long getReadTimeNanos();
 
-    long getSize();
+    long getEstimatedSize();
+
+    Slice readTail(int length)
```
Comment (electrum) [inline presto-orc/src/main/java/io/prestosql/orc/OrcDataSource.java:34]: Move this next to the other read method

File: presto-orc/src/main/java/io/prestosql/orc/OrcReader.java [hunk 12]
```diff
@@ -374,10 +387,15 @@ else if (orcType.getOrcTypeKind() == OrcTypeKind.UNION) {
...
-            throws IOException
     {
-        Slice headerMagic = source.readFully(0, MAGIC.length());
-        return MAGIC.equals(headerMagic);
+        try {
+            Slice headerMagic = source.readFully(0, MAGIC.length());
+            return MAGIC.equals(headerMagic);
+        }
+        catch (IOException e) {
+            // assume file is an ORC file
```
Comment (electrum) [inline presto-orc/src/main/java/io/prestosql/orc/OrcReader.java:396]: Why the change? Is seems the caller still handles `IOException`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/rcfile/RcFilePageSourceFactory.java [hunk 13]
```diff
@@ -134,10 +138,17 @@ else if (deserializerClassName.equals(ColumnarSerDe.class.getName())) {
...
+        RcFileDataSource dataSource;
         try {
             FileSystem fileSystem = hdfsEnvironment.getFileSystem(session.getUser(), path, configuration);
-            inputStream = hdfsEnvironment.doAs(session.getUser(), () -> fileSystem.open(path));
+            FSDataInputStream inputStream = hdfsEnvironment.doAs(session.getUser(), () -> fileSystem.open(path));
+            dataSource = new HdfsRcFileDataSource(path.toString(), inputStream, estimatedFileSize, stats);
+            if (estimatedFileSize < BUFFER_SIZE.toBytes()) {
+                byte[] buffer = new byte[toIntExact(estimatedFileSize)];
+                dataSource.readFully(0, buffer, 0, buffer.length);
+                dataSource.close();
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/rcfile/RcFilePageSourceFactory.java]: This isn't called if the read throws. I think we can wrap this in try-with-resources

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/FSDataInputStreamTail.java [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+    private final Slice tailSlice;
+    private final long fileSize;
+
+    private FSDataInputStreamTail(long fileSize, Slice tailSlice)
+    {
+        this.tailSlice = requireNonNull(tailSlice, "tailBuffer is null");
+        this.fileSize = fileSize;
+        checkArgument(fileSize >= 0, "fileSize must be >= 0, found: %s", fileSize);
+        checkArgument(tailSlice.length() <= fileSize, "length must be <= fileSize, found: %s > %s", tailSlice.length(), fileSize);
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/FSDataInputStreamTail.java:44]:
```java
checkArgument("fileSize is negative: %s", fileSize);
checkArgument(tailSlice.length() <= fileSize, "length (%s) is greater than fileSize (%s)", tailSlice.length(), fileSize);

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/FSDataInputStreamTail.java [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+    public static FSDataInputStreamTail readTail(String path, long paddedFileSize, FSDataInputStream inputStream, int length)
+            throws IOException
+    {
+        checkArgument(length >= 0 && length <= MAXIMUM_READ_LENGTH, "length too large, found: %s", length);
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/FSDataInputStreamTail.java:50]:
```
checkArgument(length >= 0, "length is negative: %s", length);
checkArgument(length <= MAXIMUM_READ_LENGTH, "length (%s) exceeds maximum (%s)", length, MAXIMUM_READ_LENGTH);

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/FSDataInputStreamTail.java [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        long position = max(paddedFileSize - MAX_SUPPORTED_PADDING_BYTES, 0);
+        long maxEOFAt = paddedFileSize + 1;
+        long startPos = inputStream.getPos();
+        try {
+            inputStream.seek(position);
+            int c;
+            while (position < maxEOFAt) {
+                c = inputStream.read();
+                if (c == -1) {
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/FSDataInputStreamTail.java:87]: Should this be `< 0` to be consistent with the above?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/util/FSDataInputStreamTail.java [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+            throw rejectInvalidFileSize(path, paddedFileSize);
+        }
+        finally {
+            inputStream.seek(startPos);
+        }
+    }
+
+    private static PrestoException rejectInvalidFileSize(String path, long reportedSize)
+    {
+        throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Incorrect fileSize %s for file %s, end of stream not reached", reportedSize, path));
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/util/FSDataInputStreamTail.java:101]:
```java
throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Incorrect file size (%s) for file (end of stream not reached): %s", reportedSize, path));

Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/util/FSDataInputStreamTail.java:101]: This should throw `IOException` since it is used by `PrestoS3FileSystem`:
```java
throw new IOException(format("Incorrect file size (%s) for file (end of stream not reached): %s", reportedSize, path));

File: presto-orc/src/main/java/io/prestosql/orc/OrcReader.java (hunk 12 above)
Comment (dain) [reply presto-orc/src/main/java/io/prestosql/orc/OrcReader.java:396]: This is only used in one place, and this the behavior that is desired there.  I just inlined this so it is easier to understand.

**Comment by dain** [comment]
@sopel39 I don't know.  @pettyjamesm wrote the commits that deal read the tail.  I dealt mostly with integrating into readers.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestOrcPageSourceMemoryTracking.java (hunk 6 above)
Comment (dain) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestOrcPageSourceMemoryTracking.java:197]: I have no idea 🤷
